File: go/internal/e2e/llm_inference_errors_e2e_test.go
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package e2e

import (
	"errors"
	"net/http"
	"strings"
	"sync"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

type llmThrowingHandler struct {
	mu               sync.Mutex
	totalCalls       int
	callsBeforeError int
}

func (h *llmThrowingHandler) OnLlmRequest(req *copilot.LlmInferenceRequest) error {
	h.mu.Lock()
	h.totalCalls++
	h.mu.Unlock()

	served, err := llmServiceNonInference(req)
	if err != nil {
		return err
	}
	if served {
		return nil
	}

	url := strings.ToLower(req.URL)
	if strings.Contains(url, "/chat/completions") || strings.Contains(url, "/responses") {
		llmDrainRequest(req)
		h.mu.Lock()
		h.callsBeforeError++
		h.mu.Unlock()
		return errors.New("synthetic-callback-transport-failure")
	}

	return llmRespondBuffered(req, 200, http.Header{"content-type": {"application/json"}}, "{}")
}

func TestLlmInferenceErrors(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	handler := &llmThrowingHandler{}
	client := newLlmClient(ctx, handler)
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
		OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
	})
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}

	// The handler raises from the inference callback; the agent layer surfaces
	// it as an error or an event rather than hanging. The assertion is loose:
	// the inference call was attempted and the runtime did not hang.
	_, sendErr := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Say OK."})
	_ = session.Disconnect()

	handler.mu.Lock()
	total := handler.totalCalls
	before := handler.callsBeforeError
	handler.mu.Unlock()

	if total == 0 {
		t.Fatal("Expected the callback to be invoked")
	}
	if before == 0 {
		t.Fatal("Expected the inference callback to be reached and raise")
	}
	if sendErr != nil && len(sendErr.Error()) == 0 {
		t.Fatal("Expected a non-empty error string when an error surfaces")
	}
}

File: java/src/test/java/com/github/copilot/LlmInferenceErrorsE2ETest.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package com.github.copilot;

import static com.github.copilot.LlmInferenceTestSupport.drainRequest;
import static com.github.copilot.LlmInferenceTestSupport.headers;
import static com.github.copilot.LlmInferenceTestSupport.newLlmClient;
import static com.github.copilot.LlmInferenceTestSupport.respondBuffered;
import static com.github.copilot.LlmInferenceTestSupport.serviceNonInference;
import static com.github.copilot.LlmInferenceTestSupport.setupCapiAuth;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Locale;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import com.github.copilot.rpc.MessageOptions;
import com.github.copilot.rpc.PermissionHandler;
import com.github.copilot.rpc.SessionConfig;

/**
 * Verifies that an exception raised from the inference callback surfaces as a
 * turn error rather than hanging the runtime.
 */
public class LlmInferenceErrorsE2ETest {

    private static E2ETestContext ctx;

    @BeforeAll
    static void setup() throws Exception {
        ctx = E2ETestContext.create();
    }

    @AfterAll
    static void teardown() throws Exception {
        if (ctx != null) {
            ctx.close();
        }
    }

    private static final class ThrowingHandler implements LlmInferenceProvider {

        private final AtomicInteger totalCalls = new AtomicInteger();
        private final AtomicInteger callsBeforeError = new AtomicInteger();

        @Override
        public void onLlmRequest(LlmInferenceRequest req) throws Exception {
            totalCalls.incrementAndGet();
            if (serviceNonInference(req)) {
                return;
            }
            String url = req.getUrl().toLowerCase(Locale.ROOT);
            if (url.contains("/chat/completions") || url.contains("/responses")) {
                drainRequest(req);
                callsBeforeError.incrementAndGet();
                throw new RuntimeException("synthetic-callback-transport-failure");
            }
            respondBuffered(req, 200, headers("content-type", "application/json"), "{}");
        }
    }

    @Test
    void surfacesHandlerErrors() throws Exception {
        setupCapiAuth(ctx);
        ThrowingHandler handler = new ThrowingHandler();

        try (CopilotClient client = newLlmClient(ctx, handler)) {
            CopilotSession session = client
                    .createSession(new SessionConfig().setOnPermissionRequest(PermissionHandler.APPROVE_ALL)).get();

            // The handler raises from the inference callback; the agent layer surfaces it
            // as an error or event rather than hanging. The assertion is loose: the
            // inference call was attempted and the runtime did not hang.
            try {
                session.sendAndWait(new MessageOptions().setPrompt("Say OK.")).get(60, TimeUnit.SECONDS);
            } catch (Exception ignored) {
                // Expected: the inference callback raised.
            }
            session.close();
        }

        assertTrue(handler.totalCalls.get() > 0, "Expected the callback to be invoked");
        assertTrue(handler.callsBeforeError.get() > 0, "Expected the inference callback to be reached and raise");
    }
}
